azure/Guidelines.md

```diff
@@ -0,0 +1,861 @@
+   "subscription": { 
+      "kind": "paid", 
+      "expiration": "2024", 
+      "invoice": "123456"
+   }
+}
+```
+Both Rectangle and Circle has common fields: ```kind```, ```fillColor```, ```lineColor```, and ```subscription```. A Rectangle also has ```x```, ```y```, ```width```, and ```length``` while a Circle has ```x```, ```y```, and ```radius```. The ```subscription``` is a nested polymorphic type. A ```free``` subscription has no additional fields and a ```paid``` subscription has ```expiration``` and ```invoice``` fields.
+
+## Common API Patterns
+
+### Performing an Action
+The REST specification is used to model the state of a resource, and is primarily intended to handle CRUD (Create, Read, Update, Delete) operations. However, many services require the ability to perform an action on a resource, e.g. getting the thumbnail of an image, sending an SMS message.
+
+> :white_check_mark: **DO** pattern your URL like this to perform an action on a resource
+**URL Pattern** 
+ ```https://.../<resource-collection>/<resource-id>/:<action>?<input parameters>```
+
+**SMS Example**
+
+ ```https://.../users/Bob/:send-sms?Text="Hello"```
+
+**Equivalent to (in C#)**
+```users["Bob"].SendSms("Hello")```
+
+> :white_check_mark: **DO** use a POST operation for any action on a resource. 
+>
+> :white_check_mark: **DO** support the Repeatability-Request-ID & Repeatability-First-Sent request headers if the action needs to be idempotent if retries occur.
+
+> :ballot_box_with_check: **YOU SHOULD** use a verb to name your action.
+
+
+### Collections
+> :white_check_mark: **DO** structure the response to a list operation as an object with a top-level array field containing the set (or subset) of resources.
+
+> :ballot_box_with_check: **YOU SHOULD** support paging today if there is ever a chance in the future that the number of items can grow to be very large. 
+> NOTE: It is a breaking change to add paging in the future
+
+> :heavy_check_mark: **YOU MAY** expose an operation that lists your resources by supporting a GET method with a URL to a resource-collection (as opposed to a resource-id).
+
+**Example Response Body** 
+```
+{
+    "value": [
+       { "id": "Item 01", "etag": "0xabc", "price": 99.95, "sizes": null },
+       { … },
+       { … },
+       { "id": "Item 99", "etag": "0xdef", "price": 59.99, "sizes": null }
+    ],
+    "nextLink": "{opaqueUrl}"
+ }
+```
+
+> :white_check_mark: **DO** include the id field and etag field (if supported) for each item as this allows the customer to modify the item in a future operation.
+> 
+> :white_check_mark: **DO** clearly document that resources may be skipped or duplicated across pages of a paginated collection unless the operation has made special provisions to prevent this (like taking a time-expiring snapshot of the collection).
+>
+> :white_check_mark: **DO** return a ```nextLink``` field with a URL that the client can GET in order to retrieve the next page of the collection.
```

Comment:
Should we clarify it should be an absolute URL callable as-is? We have precedence of people expecting people and/or SDK to rebuilt the full url with some host configuration, which has been painful in the past.

Reply:
+1. Will update